src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,768 @@
+
+* Pros
+  * If we opt-in for this we could provide compatibility with Newtonsoft which is always desired by the community.
+* Cons
+  * Quite invasive, (it affects `JsonException.Path`, `JsonSerializerOptions.IgnoreNullValues`, `JsonPropertyNameAttribute`, and Converters).
+  * This would break existing converters i.e: an array converter may expect the first token to be "[" and a preserved array starts with "{".
+    * perhaps converters are more feasible with the JSON path impl.
+  * We will now accept that an array comes in valid format when starts with a curly brace "{"; below issue is related to guard against NRE when this happens:
+    * https://github.com/dotnet/corefx/issues/41839
+
+## dojo toolkit (JavaScript framework)
+https://dojotoolkit.org/reference-guide/1.10/dojox/json/ref.html
+
+Similar: https://www.npmjs.com/package/json-cyclic 
+
+* id-based (ignore this approach since is the same the one of Newtonsoft.Json)
+* path-based
+  * "\#" denotes the root of the object and then uses semantics inspired in JSONPath.
+  * It does not uses `$id` nor `$values` metadata, therefore, everything can be referenced.
+  * Pros
+    * It looks cleaner. 
+    * Only disruptive (weird) edge case would be a reference to an array i.e: { "MyArray": { "$ref": "#manager.subordinates" } }.
+  * Cons
+    * Path value will become too long on very deep objects.
+    * Storing all the complex types could become very expensive, are we going to store also primitive types?
+    * This would break existing converters when handling reference to an array.
+    * Not compatible with Newtonsoft.Json.
+  
+## flatted (JavaScript module) (probably not worth it)
+https://github.com/WebReflection/flatted
+
+* While stringifying, all Objects, including Arrays, and strings, are flattened out and replaced as unique index.
+* Once parsed, all indexes will be replaced through the flattened collection.
+* It has 23M downloads per month.
+* Every single value (primitive and complex) is preserved.
+* Cons:
+  * It does not look like JSON anymore.
+
+
+## Jackson (Java)
+https://www.baeldung.com/jackson-bidirectional-relationships-and-infinite-recursion
+
+* Let you annotate your class with @JsonIdentityInfo where you can define a class property that will be used to further represent the object.
+
+## golang 
+
+* (https://go-review.googlesource.com/c/go/+/187920/), The fix is about detecting circular references after a threshold of 1,000 and throw when found in order to prevent a non-recoverable Stack Overflow.
+
+# Proposal
+
+```cs
+namespace System.Text.Json.Serialization
+{
+    /// <summary>
+    /// This class defines the various ways the <see cref="JsonSerializer"/> can deal with references on Serialization and Deserialization.
+    /// </summary>
+    public sealed class ReferenceHandling
+    {
+        public static ReferenceHandling Default { get; }
+        // TODO: decide if we keep or remove this option.
+        public static ReferenceHandling Ignore { get; }
```

Comment:
nit: consider moving this after `Preserve` to it is clear that we may not expose it.